### PR TITLE
Allows shells that requiring anchoring to take power from APCs

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request

Allows anchored shells to take power from APCs. If they take too much in a short period of time, they blow up and automatically eject the circuit to prevent someone from attempting to drain the grid.
Doors shells have been made to require anchoring but this'll have no player facing changes because they cannot be unanchored anyways.

## Why It's Good For The Game

Door shells require power to function and also require a separate cell to function too. This also improves the server shell as it makes it more viable as something that can handle heavy processing without needing to worry about recharging it.

## Changelog

:cl:
expansion: Shells that require anchoring to work now take their power from the room's APC rather than their cell. They can still use their cell as a backup battery. If they take too much power in a short period of time, they'll blow up and eject the circuit. You can examine the shell to see if it requires a cell to function.
/:cl:

### CODE REVIEW

**Comment by TheBonded** [comment]
All that's left now is to give USB compatibility to server shells so you can have fully automated USB circuits that take power directly from the grid

**Comment by Watermelon914** [comment]
> All that's left now is to give USB compatibility to server shells so you can have fully automated USB circuits that take power directly from the grid

they already use usbs

**Comment by TheBonded** [comment]
> 
> 
> > All that's left now is to give USB compatibility to server shells so you can have fully automated USB circuits that take power directly from the grid
> 
> they already are able to use usbs

oh, huh, thought it was only bot shells

File: code/datums/components/shell.dm [hunk 1]
```diff
@@ -14,8 +14,19 @@
...
 
+	/// Whether the shell is locked or not
 	var/locked = FALSE
 
+	// The variables below are used only for anchored shells
+	/// The amount of power used in the last minute
+	var/power_used_in_minute = 0
+
+	/// The last time power was taken from the grid.
+	var/power_used_last = 0
```
Comment (Mothblocks) [inline code/datums/components/shell.dm:25]: Cooldown define

File: code/datums/components/shell.dm [hunk 2]
```diff
@@ -256,6 +271,22 @@
...
 
+/datum/component/shell/proc/override_power_usage(datum/source, power_to_use)
+	SIGNAL_HANDLER
+	if(world.time - power_used_last > 1 MINUTES)
+		power_used_in_minute = 0
+
+	var/area/location = get_area(parent)
+	if(location.powered(AREA_USAGE_EQUIP))
+		if(power_used_in_minute > max_power_use_in_minute)
+			explosion(parent, 0, 0, 1, explosion_cause = attached_circuit)
```
Comment (Mothblocks) [inline code/datums/components/shell.dm:282]: Named vars for 0, 0, 1.

File: code/datums/components/shell.dm [hunk 3]
```diff
@@ -256,6 +271,22 @@
 		source.balloon_alert(user, "it won't fit!")
 		return COMPONENT_CANCEL_ADD_COMPONENT
 
+/datum/component/shell/proc/override_power_usage(datum/source, power_to_use)
+	SIGNAL_HANDLER
+	if(world.time - power_used_last > 1 MINUTES)
+		power_used_in_minute = 0
+
+	var/area/location = get_area(parent)
+	if(location.powered(AREA_USAGE_EQUIP))
```
Comment (Mothblocks) [inline code/datums/components/shell.dm:280]: Early-return.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Pull request merged**